# Datasets
![–î–∞—Ç–∞—Å–µ—Ç—ã, –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–Ω—ã–µ –ø—Ä–∏ –æ–±—É—á–µ–Ω–∏–∏ CRISPR-BERT](/images/datasets_info.bmp)
*Datasets used for CRISPR-BERT training* (*[Luo et al., 2024](https://www.sciencedirect.com/science/article/pii/S0010482524000167)*)

## K562 and HEK293t
The off-target profile dataset contained two different cell types: 293-related cell lines (18 sgRNAs) and K562 t (12 sgRNAs) [6, 22, 24,25,26,27, 44]. For all 30 sgRNAs, we obtained ~‚Äâ160,000 possible loci across the whole genome using bowtie2 [51], with a maximum of six nucleotide mismatches. The whole dataset was highly unbalanced, and nearly 1/250 loci was identified as an off-target site (one mismatch, 4; two mismatches, 31; three mismatches, 121; four mismatches, 236; five mismatches, 174; six mismatches, 75) with various whole genome off-target detection techniques [22,23,24,25,26,27] (Additional files 6 and 7). For the classification model, the off-target sites are labeled as ‚Äú1‚Äù and the others are labeled as ‚Äú0‚Äù. For the regression model, the off-target sites are labeled and normalized with the targeting cleavage frequency (indel frequency) detected by different off-target detection assays. (*Source: [G. Chuai et al. (2018) DeepCRISPR: optimized CRISPR guide RNA design by deep learning](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1459-4)*)

Sample of K562 data (20319 rows):
| target | potential_off_target | is_off_target |
|----------|----------|----------|
| AAATGAGAAGAAGAGGCACAGGG | GCATGAGAAGAAGAGACATAGCC | 0 |
| AAATGAGAAGAAGAGGCACAGGG | GAAGAAGAAGAAGAGGAAGAGGA | 0 |
| CCTGCCTCCGCTCTACTCACTGG | CCTGGCTCCTCTCTCCTCACTGG | 1 |

*Dataset source: [S.W. Cho et al. (2014) Analysis of off-target effects of CRISPR/Cas-derived RNA-guided endonucleases and nickases](https://pmc.ncbi.nlm.nih.gov/articles/PMC3875854/)*

## Table_S8_machine_learning_input

26248 rows
The genomic sequence extending from 22 bases 5‚Ä≤ of the beginning of the PAM to 1 base 3‚Ä≤ of the end of the PAM

1. *key* - —É–Ω–∏–∫–∞–ª—å–Ω–æ–µ –∑–Ω–∞—á–µ–Ω–∏–µ
2. *perfect_match_sgRNA* - –≥–∏–¥–†–ù–ö, –ø–æ–ª–Ω–æ—Å—Ç—å—é –∫–æ–º–ø–ª–µ–º–µ–Ω—Ç–∞—Ä–Ω–∞—è —Ç–∞—Ä–≥–µ—Ç–Ω–æ–º—É —É—á–∞—Å—Ç–∫—É –î–ù–ö (perfectly matched sgRNA, PMsgRNA)
3. *gene* - –Ω–∞–∑–≤–∞–Ω–∏–µ –≥–µ–Ω–∞
4. *sgRNA_sequence* - –≥–∏–¥–†–ù–ö —Å –æ–¥–Ω–∏–º –Ω–µ—Å–æ–≤–ø–∞–¥–µ–Ω–∏–µ–º –∫ —Ç–∞—Ä–≥–µ—Ç–Ω–æ–º—É —É—á–∞—Å—Ç–∫—É (mismatched sgRNA, MMsgRNA)
5. *mismatch_position* - –ø–æ–∑–∏—Ü–∏—è –Ω–µ—Å–æ–≤–ø–∞–¥–µ–Ω–∏—è 
6. *new_pairing* - —Å–æ–¥–µ—Ä–∂–∞–Ω–∏–µ –Ω–µ—Å–æ–≤–ø–∞–¥–µ–Ω–∏—è. –ù–∞–ø—Ä–∏–º–µ—Ä, *rA:dC*	–∑–Ω–∞—á–∏—Ç, —á—Ç–æ –Ω—É–∫–ª–µ–æ—Ç–∏–¥ G (–∫–æ–º–ø–ª–µ–º–µ–Ω—Ç–∞—Ä–Ω—ã–π C) –≤ –†–ù–ö –∑–∞–º–µ–Ω–µ–Ω –Ω–∞ A.
7. *K562* - –±—ã–ª–∏ –ª–∏ –ø—Ä–æ–≤–µ–¥–µ–Ω—ã —ç–∫—Å–ø–µ—Ä–∏–º–µ–Ω—Ç—ã –Ω–∞ –∫–ª–µ—Ç–∫–∞—Ö –ª–∏–Ω–∏–∏ K562
8. *Jurkat* - –±—ã–ª–∏ –ª–∏ –ø—Ä–æ–≤–µ–¥–µ–Ω—ã —ç–∫—Å–ø–µ—Ä–∏–º–µ–Ω—Ç—ã –Ω–∞ –∫–ª–µ—Ç–∫–∞—Ö –ª–∏–Ω–∏–∏ Jurkat
9. *mean_relative_gamma* - –∞–∫—Ç–∏–≤–Ω–æ—Å—Ç—å MMsgRNA –æ—Ç–Ω–æ—Å–∏—Ç–µ–ª—å–Ω–æ PMsgRNA

*Dataset source: [M. Jost et al. (2020) Titrating gene expression using libraries of systematically attenuated CRISPR guide RNAs](https://pmc.ncbi.nlm.nih.gov/articles/PMC7065968/)*

### –û–ø–∏—Å–∞–Ω–∏–µ —Ü–µ–ª–µ–≤–æ–π –≤–µ–ª–∏—á–∏–Ω—ã mean relative gamma
**Gamma (Œ≥)** ‚Äî —ç—Ç–æ –º–µ—Ä–∞ –≤–ª–∏—è–Ω–∏—è –∫–æ–Ω–∫—Ä–µ—Ç–Ω–æ–π sgRNA –Ω–∞ —Ä–æ—Å—Ç –∫–ª–µ—Ç–æ–∫ –≤ –ø—É–ª–µ (–ø–æ–ø—É–ª—è—Ü–∏–∏). –û–ø—Ä–µ–¥–µ–ª—è–µ—Ç—Å—è –ø–æ —Å–ª–µ–¥—É—é—â–µ–π —Ñ–æ—Ä–º—É–ª–µ:

<img src="./images/gamma_formula.png" style="height: 80px;">

ùëÅ initial,sgRNA, ùëÅ final,sgRNA ‚Äî –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ –∫–ª–µ—Ç–æ–∫ —Å –æ–ø—Ä–µ–¥–µ–ª–µ–Ω–Ω–æ–π sgRNA –≤ –Ω–∞—á–∞–ª–µ –∏ –∫–æ–Ω—Ü–µ —ç–∫—Å–ø–µ—Ä–∏–º–µ–Ω—Ç–∞.

–ê–Ω–∞–ª–æ–≥–∏—á–Ω–æ –¥–ª—è non-targeting sgRNA (–∫–æ–Ω—Ç—Ä–æ–ª—å–Ω—ã–µ sgRNA, –∫–æ—Ç–æ—Ä—ã–µ –Ω–µ –¥–µ–π—Å—Ç–≤—É—é—Ç –Ω–∞ —Ü–µ–ª–µ–≤–æ–π –≥–µ–Ω).

–¢–∞–∫–∏–º –æ–±—Ä–∞–∑–æ–º, ùõæ –æ—Ç—Ä–∞–∂–∞–µ—Ç, –Ω–∞—Å–∫–æ–ª—å–∫–æ –∫–ª–µ—Ç–∫–∞ —Å –∫–æ–Ω–∫—Ä–µ—Ç–Ω–æ–π sgRNA —Ä–∞—Å—Ç–µ—Ç –ª—É—á—à–µ (ùõæ > 0) –∏–ª–∏ —Ö—É–∂–µ (ùõæ < 0), —á–µ–º –∫–ª–µ—Ç–∫–∏ —Å –∫–æ–Ω—Ç—Ä–æ–ª—å–Ω—ã–º–∏ (non-targeting) sgRNA. –ß–µ–º –±–æ–ª—å—à–µ –æ—Ç—Ä–∏—Ü–∞—Ç–µ–ª—å–Ω–æ–µ ùõæ, —Ç–µ–º —Å–∏–ª—å–Ω–µ–µ –Ω–µ–≥–∞—Ç–∏–≤–Ω—ã–π —ç—Ñ—Ñ–µ–∫—Ç sgRNA –Ω–∞ —Ä–æ—Å—Ç –∫–ª–µ—Ç–æ–∫.

**Relative gamma** ‚Äî —ç—Ç–æ –Ω–æ—Ä–º–∞–ª–∏–∑–æ–≤–∞–Ω–Ω–æ–µ –∑–Ω–∞—á–µ–Ω–∏–µ Œ≥, –ø–æ–∫–∞–∑—ã–≤–∞—é—â–µ–µ –∞–∫—Ç–∏–≤–Ω–æ—Å—Ç—å sgRNA —Å –Ω–µ—Å–æ–≤–ø–∞–¥–µ–Ω–∏—è–º–∏ (mis-matched sgRNA) –æ—Ç–Ω–æ—Å–∏—Ç–µ–ª—å–Ω–æ –∏–¥–µ–∞–ª—å–Ω–æ —Å–æ–≤–ø–∞–¥–∞—é—â–µ–π sgRNA (perfect match sgRNA):

<img src="./images/relative_gamma_formula.png" style="height: 70px;">

sgRNA_variant ‚Äî —ç—Ç–æ sgRNA —Å –Ω–µ—Å–æ–≤–ø–∞–¥–µ–Ω–∏–µ–º (–º–∏—Å–º–∞—Ç—á–µ–º).

sgRNA_perfect ‚Äî –∏–¥–µ–∞–ª—å–Ω–æ —Å–æ–≤–ø–∞–¥–∞—é—â–∞—è sgRNA (–ø–æ–ª–Ω–æ–µ —Å–æ–≤–ø–∞–¥–µ–Ω–∏–µ —Å —Ü–µ–ª–µ–≤–æ–π –ø–æ—Å–ª–µ–¥–æ–≤–∞—Ç–µ–ª—å–Ω–æ—Å—Ç—å—é).

- –ó–Ω–∞—á–µ–Ω–∏–µ 1 –æ–∑–Ω–∞—á–∞–µ—Ç, —á—Ç–æ sgRNA —Å –Ω–µ—Å–æ–≤–ø–∞–¥–µ–Ω–∏–µ–º –¥–µ–π—Å—Ç–≤—É–µ—Ç —Ç–∞–∫ –∂–µ, –∫–∞–∫ –∏–¥–µ–∞–ª—å–Ω–æ —Å–æ–≤–ø–∞–¥–∞—é—â–∞—è sgRNA.
- –ó–Ω–∞—á–µ–Ω–∏–µ 0 –æ–∑–Ω–∞—á–∞–µ—Ç, —á—Ç–æ sgRNA —Å –Ω–µ—Å–æ–≤–ø–∞–¥–µ–Ω–∏–µ–º –Ω–µ –∏–º–µ–µ—Ç –≤–ª–∏—è–Ω–∏—è –Ω–∞ —Ä–æ—Å—Ç (—Ç–æ –µ—Å—Ç—å –µ—ë –∞–∫—Ç–∏–≤–Ω–æ—Å—Ç—å –æ—Ç—Å—É—Ç—Å—Ç–≤—É–µ—Ç –∏–ª–∏ —Ä–∞–≤–Ω–∞ –∫–æ–Ω—Ç—Ä–æ–ª—å–Ω–æ–π).

–ü–æ–∫–∞–∑–∞—Ç–µ–ª—å **Mean Relative Gamma** –ø–æ–ª—É—á–∞–µ—Ç—Å—è –ø—É—Ç—ë–º —É—Å—Ä–µ–¥–Ω–µ–Ω–∏—è relative gamma –ø–æ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–∞–º –Ω–µ—Å–∫–æ–ª—å–∫–∏—Ö —ç–∫—Å–ø–µ—Ä–∏–º–µ–Ω—Ç–æ–≤ (–≤ —Å—Ç–∞—Ç—å–µ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–ª–∏ –∫–ª–µ—Ç–∫–∏ K562 –∏ Jurkat):

<img src="./images/mean_relative_gamma_formula.png" style="height: 50px;">

–≠—Ç–æ —É—Å—Ä–µ–¥–Ω–µ–Ω–∏–µ –¥–µ–ª–∞–µ—Ç –º–µ—Ç—Ä–∏–∫—É –±–æ–ª–µ–µ —É—Å—Ç–æ–π—á–∏–≤–æ–π –∏ –º–µ–Ω–µ–µ —á—É–≤—Å—Ç–≤–∏—Ç–µ–ª—å–Ω–æ–π –∫ —à—É–º—É.

| –î–∏–∞–ø–∞–∑–æ–Ω –∑–Ω–∞—á–µ–Ω–∏–π | –ò–Ω—Ç–µ—Ä–ø—Ä–µ—Ç–∞—Ü–∏—è |
|----------|----------|
| <0 | –ù–µ—Å–æ–≤–ø–∞–¥–∞—é—â–∞—è sgRNA –æ–∫–∞–∑—ã–≤–∞–µ—Ç –ø—Ä–æ—Ç–∏–≤–æ–ø–æ–ª–æ–∂–Ω—ã–π —ç—Ñ—Ñ–µ–∫—Ç –Ω–∞ —Ä–æ—Å—Ç –∫–ª–µ—Ç–æ–∫, —á–µ–º –∏–¥–µ–∞–ª—å–Ω–æ —Å–æ–≤–ø–∞–¥–∞—é—â–∞—è (–Ω–∞–ø—Ä–∏–º–µ—Ä, –≤–º–µ—Å—Ç–æ –∏–Ω–≥–∏–±–∏—Ä–æ–≤–∞–Ω–∏—è, —Å—Ç–∏–º—É–ª–∏—Ä—É–µ—Ç —Ä–æ—Å—Ç). |
| 0 | –ü–æ–ª–Ω–æ—Å—Ç—å—é –Ω–µ–∞–∫—Ç–∏–≤–Ω–∞—è sgRNA —Å –Ω–µ—Å–æ–≤–ø–∞–¥–µ–Ω–∏–µ–º (–Ω–µ –≤–ª–∏—è–µ—Ç –Ω–∞ —Ä–æ—Å—Ç) |
| 0,1-0,9 | –ß–∞—Å—Ç–∏—á–Ω–∞—è –∞–∫—Ç–∏–≤–Ω–æ—Å—Ç—å sgRNA (—É–º–µ—Ä–µ–Ω–Ω–æ–µ –ø–æ–¥–∞–≤–ª–µ–Ω–∏–µ —ç–∫—Å–ø—Ä–µ—Å—Å–∏–∏ –≥–µ–Ω–∞ –∏ –≤–ª–∏—è–Ω–∏—è –Ω–∞ —Ä–æ—Å—Ç –∫–ª–µ—Ç–æ–∫). |
| ~1 | 	–ù–µ—Å–æ–≤–ø–∞–¥–∞—é—â–∞—è sgRNA –ø–æ—á—Ç–∏ —Ç–∞–∫ –∂–µ —ç—Ñ—Ñ–µ–∫—Ç–∏–≤–Ω–∞, –∫–∞–∫ –∏–¥–µ–∞–ª—å–Ω–æ —Å–æ–≤–ø–∞–¥–∞—é—â–∞—è. |
| >1 | –ù–µ—Å–æ–≤–ø–∞–¥–∞—é—â–∞—è sgRNA –Ω–µ–æ–∂–∏–¥–∞–Ω–Ω–æ –±–æ–ª–µ–µ —ç—Ñ—Ñ–µ–∫—Ç–∏–≤–Ω–∞ –≤ –ø–æ–¥–∞–≤–ª–µ–Ω–∏–∏ —ç–∫—Å–ø—Ä–µ—Å—Å–∏–∏ –≥–µ–Ω–∞, —á–µ–º –∏–¥–µ–∞–ª—å–Ω–æ —Å–æ–≤–ø–∞–¥–∞—é—â–∞—è –ò–õ–ò perfectRNA —É—Å–∏–ª–∏–ª–∞ —Ä–æ—Å—Ç –∫–ª–µ—Ç–æ–∫, –∞ mismatched —É—Å–∏–ª–∏–ª–∞ –µ–≥–æ –µ—â–µ –±–æ–ª—å—à–µ |

## 4 cell lines

Made from the datasets used for the study of sgRNA on-target efficacy prediction. Includes data for four cell lines: *hct116, hek293t, hela, hl60*

Sample (16749 rows):
| Chromosome | Start | End | Strand | sgRNA | Normalized efficacy | cell_line |
|----------|----------|----------|----------|----------|----------|----------|
| chr17 | 33469132 | 33469154 | - | CTTGCTCGCGCAGGACGAGGCGG | 0.16401994504229 | hct116 |
| chr1 | 12316394 | 12316416 | - | ACTCTCCATGCTCCTGGCCATGG | 0.00942528735632599 | hela |

## Data augmented dataset for on-target prediction
This dataset was generated from ~15,000 sgRNAs across 1071 genes with known knockout efficacies in a data augmentation manner (see the [‚ÄúOn-target data sources‚Äù](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1459-4#Sec24) section).

Previous studies have shown that PAM-distal region has a high tolerance for sequence mismatches (Kim et al., 2016; Kleinstiver et al., 2016). To be specific, gRNAs with two mismatches in the first two positions from the 5‚Äô end has little influence on cleavage efficiency (Doench et al., 2014; Doench et al., 2016). Inspired by these studies, Chuai et al. applied a data augmentation procedure by changing each gRNA into a new one with two mismatches in the PAM distal region (Chuai et al., 2018). Consequently, a 23-nt gRNA sequence can be expanded into 16 gRNAs with identical cleavage efficacy. The augmented dataset was generated from  15,000 gRNAs with known on-target cleavage efficacy. By adopting this data augmentation strategy, they obtained 180,512 non-redundant gRNAs. Each observation in the data contains a 23-nt gRNA sequence and its corresponding cleavage efficiency (*Source: [A. F. Abbasi et al. (2025) Transitioning from wet lab to artificial intelligence: a systematic review of AI predictors in CRISPR](https://pmc.ncbi.nlm.nih.gov/articles/PMC11796103/)*)

Sample (180513 rows):

| sgRNA | Efficacy | 
|----------|----------|
| AAACCGCCTCCAAGTCGCCGAGG | 1 |
| AAGTGGCTTCCGTGGTGGCCGGG | 0 |

*Dataset source: [G. Chuai et al. (2018) DeepCRISPR: optimized CRISPR guide RNA design by deep learning](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1459-4)*

# Embeddings

## –ü–æ–ø—Ä–æ–±–æ–≤–∞—Ç—å
mRNABERT, RNABERT, DNABERT, DNABERT-2, RNAGPT, HyenaDNA, IsoFormer 

## SeQuant
[GitHub](https://github.com/GenerativeMolMachines/SeQuant)

-----------------------------------------------------------------------

## –û–ø–∏—Å–∞–Ω–∏–µ –ø—Ä–æ–µ–∫—Ç–∞

–≠—Ç–æ—Ç –ø—Ä–æ–µ–∫—Ç –ø—Ä–µ–¥—Å—Ç–∞–≤–ª—è–µ—Ç —Å–æ–±–æ–π –ø–∞–π–ø–ª–∞–π–Ω –¥–ª—è –æ–±—Ä–∞–±–æ—Ç–∫–∏ –¥–∞–Ω–Ω—ã—Ö. –û—Å–Ω–æ–≤–Ω–∞—è —Ü–µ–ª—å –ø–∞–π–ø–ª–∞–π–Ω–∞ ‚Äî –∑–∞–≥—Ä—É–∑–∫–∞, –≤–∞–ª–∏–¥–∞—Ü–∏—è, —Ç—Ä–∞–Ω—Å—Ñ–æ—Ä–º–∞—Ü–∏—è –¥–∞–Ω–Ω—ã—Ö –∏ –∏—Ö —Å–æ—Ö—Ä–∞–Ω–µ–Ω–∏–µ –≤ –±–∞–∑—É –¥–∞–Ω–Ω—ã—Ö SQLite. –ü—Ä–æ—Ü–µ—Å—Å –≤–∫–ª—é—á–∞–µ—Ç —Å–ª–µ–¥—É—é—â–∏–µ —à–∞–≥–∏:

1. –°–∫–∞—á–∏–≤–∞–Ω–∏–µ –∏—Å—Ö–æ–¥–Ω–æ–≥–æ —Ñ–∞–π–ª–∞ –¥–∞–Ω–Ω—ã—Ö.
2. –ü—Ä–µ–æ–±—Ä–∞–∑–æ–≤–∞–Ω–∏–µ —Ç–µ–∫—Å—Ç–∞ –≤ DataFrame.
3. –ó–∞–≥—Ä—É–∑–∫–∞ —Å—ã—Ä—ã—Ö –¥–∞–Ω–Ω—ã—Ö –≤ –±–∞–∑—É –¥–∞–Ω–Ω—ã—Ö.
4. –í–∞–ª–∏–¥–∞—Ü–∏—è –∏ –¥–æ–±–∞–≤–ª–µ–Ω–∏–µ –Ω–æ–≤—ã—Ö –ø—Ä–∏–∑–Ω–∞–∫–æ–≤.
5. –°–æ—Ö—Ä–∞–Ω–µ–Ω–∏–µ –æ—á–∏—â–µ–Ω–Ω—ã—Ö –¥–∞–Ω–Ω—ã—Ö –≤ –æ—Ç–¥–µ–ª—å–Ω—É—é —Ç–∞–±–ª–∏—Ü—É –ë–î.

–†–∞–±–æ—Ç–∞ –≤—ã–ø–æ–ª–Ω—è–ª–∞—Å—å –∫–∞–∫ —á–∞—Å—Ç—å –ø—Ä–æ–µ–∫—Ç–∞ "–°–æ–∑–¥–∞–Ω–∏–µ–º –∏–Ω—Å—Ç—Ä—É–º–µ–Ω—Ç–∞ –¥–ª—è –æ—Ü–µ–Ω–∫–∏ —ç—Ñ—Ñ–µ–∫—Ç–∏–≤–Ω–æ—Å—Ç–∏ –Ω–∞–ø—Ä–∞–≤–ª—è—é—â–∏—Ö –†–ù–ö –≤ CRISPR-Cas —Å–∏—Å—Ç–µ–º–∞—Ö" (–±–æ–ª–µ–µ –ø–æ–¥—Ä–æ–±–Ω–æ–µ –æ–ø–∏—Å–∞–Ω–∏–µ –ø—Ä–æ–µ–∫—Ç–∞ –∏ –µ–≥–æ —Ü–µ–Ω–Ω–æ—Å—Ç—å –º–æ–∂–Ω–æ –ø–æ—Å–º–æ—Ç—Ä–µ—Ç—å –≤ –ø—Ä–µ–∑–µ–Ω—Ç–∞—Ü–∏–∏). –í –¥–∞–ª—å–Ω–µ–π—à–µ–º —ç—Ç–∏ –¥–∞–Ω–Ω—ã–µ –±—É–¥—É—Ç –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω—ã –¥–ª—è –æ–±—É—á–µ–Ω–∏—è –∞–ª–≥–æ—Ä–∏—Ç–º–æ–≤ –ú–û –∏/–∏–ª–∏ –Ω–µ–π—Ä–æ—Å–µ—Ç–µ–π –¥–ª—è –ø—Ä–µ–¥—Å–∫–∞–∑–∞–Ω–∏—è —Ç–∞—Ä–≥–µ—Ç–Ω–æ–π –≤–µ–ª–∏—á–∏–Ω—ã. 

## –°–æ–∑–¥–∞–Ω–∏–µ –Ω–æ–≤—ã—Ö –ø—Ä–∏–∑–Ω–∞–∫–æ–≤
–í –¥–∞—Ç–∞—Å–µ—Ç –¥–æ–±–∞–≤–ª–µ–Ω—ã –ø—è—Ç—å –Ω–æ–≤—ã—Ö –ø—Ä–∏–∑–Ω–∞–∫–æ–≤, —Ç—Ä–∏ –∏–∑ –Ω–∏—Ö - —ç—Ç–æ –∑–∞–∫–æ–¥–∏—Ä–æ–≤–∞–Ω–Ω—ã–µ —Å –ø–æ–º–æ—â—å—é —Ä–∞–∑–ª–∏—á–Ω—ã—Ö —Å—Ö–µ–º –∫–æ–¥–∏—Ä–æ–≤–∞–Ω–∏—è –ø–∞—Ä—ã –†–ù–ö-–î–ù–ö (–¥–ª–∏–Ω—ã N): 
1. *encoded_or* - –º–∞—Ç—Ä–∏—Ü–∞ 4xN. –°–Ω–∞—á–∞–ª–∞ –î–ù–ö –∏ –†–ù–ö one-hot –∫–æ–¥–∏—Ä—É—é—Ç—Å—è –ø–æ —á–µ—Ç—ã—Ä–µ–º –∫–∞–Ω–∞–ª–∞–º (ATGC), –∑–∞—Ç–µ–º –º–µ–∂–¥—É –Ω–∏–º–∏ –≤—ã–ø–æ–ª–Ω—è–µ—Ç—Å—è –æ–ø–µ—Ä–∞—Ü–∏—è –ò–õ–ò. –ü–æ–¥—Ä–æ–±–Ω–µ–µ –≤ [—Å—Ç–∞—Ç—å–µ](https://pmc.ncbi.nlm.nih.gov/articles/PMC6129261/)
2. *encoded_stacked* - –º–∞—Ç—Ä–∏—Ü–∞ 8xN. –°–Ω–∞—á–∞–ª–∞ –î–ù–ö –∏ –†–ù–ö one-hot –∫–æ–¥–∏—Ä—É—é—Ç—Å—è –ø–æ —á–µ—Ç—ã—Ä–µ–º –∫–∞–Ω–∞–ª–∞–º (ATGC), –∑–∞—Ç–µ–º –æ–Ω–∏ "—Å—Ç–µ–∫–∞—é—Ç—Å—è". –ü–æ–¥—Ä–æ–±–Ω–µ–µ –≤ [—Å—Ç–∞—Ç—å–µ](https://academic.oup.com/bioinformatics/article/37/16/2299/6143578)
3. *encoded_7channels* - –º–∞—Ç—Ä–∏—Ü–∞ 7xN. –ü–æ–∑–≤–æ–ª—è–µ—Ç —Ä–∞–±–æ—Ç–∞—Ç—å —Å –ª–∏—à–Ω–∏–º–∏ –∏–ª–∏ –Ω–µ–¥–æ—Å—Ç–∞—é—â–∏–º–∏ –Ω—É–∫–ª–µ–æ—Ç–∏–¥–∞–º–∏ –≤ –†–ù–ö (–æ—Ç–Ω–æ—Å–∏—Ç–µ–ª—å–Ω–æ —Ç–∞—Ä–≥–µ—Ç–Ω–æ–≥–æ —É—á–∞—Å—Ç–∫–∞ –î–ù–ö), –∞ —Ç–∞–∫–∂–µ —É—á–∏—Ç—ã–≤–∞–µ—Ç —Ä–∞–∑–ª–∏—á–Ω—ã–µ —Ñ—É–Ω–∫—Ü–∏–æ–Ω–∞–ª—å–Ω—ã–µ —É—á–∞—Å—Ç–∫–∏ –†–ù–ö (PAM –∏–ª–∏ —Å–æ–±—Å—Ç–≤–µ–Ω–Ω–æ –Ω–∞–ø—Ä–∞–≤–ª—è—é—â–∞—è –†–ù–ö). –ü–æ–¥—Ä–æ–±–Ω–µ–µ –≤ [—Å—Ç–∞—Ç—å–µ](https://www.sciencedirect.com/science/article/pii/S2001037022000137)
4. *gc_content* (–≥—É–∞–Ω–∏–Ω-—Ü–∏—Ç–æ–∑–∏–Ω–æ–≤—ã–π —Å–æ—Å—Ç–∞–≤, –ì–¶-—Å–æ—Å—Ç–∞–≤) - –¥–æ–ª—è –≥—É–∞–Ω–∏–Ω–∞ (G) –∏ —Ü–∏—Ç–æ–∑–∏–Ω–∞ (C) —Å—Ä–µ–¥–∏ –≤—Å–µ—Ö –Ω—É–∫–ª–µ–æ—Ç–∏–¥–æ–≤ –ø–æ—Å–ª–µ–¥–æ–≤–∞—Ç–µ–ª—å–Ω–æ—Å—Ç–∏. –ì–¶-—Å–æ—Å—Ç–∞–≤ –Ω–∞–ø—Ä–∞–≤–ª—è—é—â–µ–π –†–ù–ö –≤–ª–∏—è–µ—Ç –Ω–∞ –µ–µ —ç—Ñ—Ñ–µ–∫—Ç–∏–≤–Ω–æ—Å—Ç—å, –æ–ø—Ç–∏–º–∞–ª—å–Ω—ã–º —Å—á–∏—Ç–∞–µ—Ç—Å—è –∑–Ω–∞—á–µ–Ω–∏–µ 40-60% ([V. Konstantakos et al. 2022, CRISPR‚ÄìCas9 gRNA efficiency prediction: an overview of predictive tools and the role of deep learning ](https://academic.oup.com/nar/article/50/7/3616/6555429))
5. *pam* (protospacer adjacent motif; –º–æ—Ç–∏–≤, —Å–º–µ–∂–Ω—ã–π —Å –ø—Ä–æ—Ç–æ—Å–ø–µ–π—Å–µ—Ä–æ–º) - –∫–æ—Ä–æ—Ç–∫–∞—è –ø–æ—Å–ª–µ–¥–æ–≤–∞—Ç–µ–ª—å–Ω–æ—Å—Ç—å –î–ù–ö (–æ–±—ã—á–Ω–æ –¥–ª–∏–Ω–æ–π 2-6 –ø–∞—Ä –æ—Å–Ω–æ–≤–∞–Ω–∏–π), –æ–Ω–∞ —Å–ª–µ–¥—É–µ—Ç –∑–∞ —Ç–∞—Ä–≥–µ—Ç–Ω—ã–º —É—á–∞—Å—Ç–∫–æ–º –î–ù–ö –∏ —Å–ª—É–∂–∏—Ç "—Å–∏–≥–Ω–∞–ª–æ–º" –∫ —Ä–∞–∑—Ä–µ–∑—É. –î–∞–Ω–Ω—ã–µ –≤ –Ω–∞—à–µ–º –¥–∞—Ç–∞—Å–µ—Ç–µ —Å–æ–±—Ä–∞–Ω—ã –¥–ª—è —ç–Ω–∑–∏–º–∞ Sp.Cas9, –∞ –µ–º—É —Å–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤—É–µ—Ç PAM –≤–∏–¥–∞ 5'-NGG-3', –≥–¥–µ N-–ª—é–±–æ–π –Ω—É–∫–ª–µ–æ—Ç–∏–¥. [–ü–æ–¥—Ä–æ–±–Ω–µ–µ](https://www.addgene.org/guides/crispr/)

## EDA
–ò—Å—Å–ª–µ–¥–æ–≤–∞—Ç–µ–ª—å—Å–∫–∏–π –∞–Ω–∞–ª–∏–∑ –¥–∞–Ω–Ω—ã—Ö –ø—Ä–µ–¥—Å—Ç–∞–≤–ª–µ–Ω –≤ —Ñ–∞–π–ª–µ EDA.ipynb

## –ú–µ—Ç—Ä–∏–∫–∏ –∫–∞—á–µ—Å—Ç–≤–∞ –¥–∞–Ω–Ω—ã—Ö
### –°–æ–¥–µ—Ä–∂–∞–Ω–∏–µ –¥–∞–Ω–Ω—ã—Ö

**–í–∞–ª–∏–¥–Ω–æ—Å—Ç—å** - —Å–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤–∏–µ –º–Ω–æ–≥–æ—á–∏—Å–ª–µ–Ω–Ω—ã–º –∞—Ç—Ä–∏–±—É—Ç–∞–º, —Å–≤—è–∑–∞–Ω–Ω—ã—Ö —Å —ç–ª–µ–º–µ–Ω—Ç–æ–º –¥–∞–Ω–Ω—ã—Ö: —Ç–∏–ø, —Ç–æ—á–Ω–æ—Å—Ç—å, —Ñ–æ—Ä–º–∞—Ç, –¥–∏–∞–ø–∞–∑–æ–Ω—ã –¥–æ–ø—É—Å—Ç–∏–º—ã—Ö –∑–Ω–∞—á–µ–Ω–∏–π –∏ —Ç–∞–∫ –¥–∞–ª–µ–µ.

1. *key* - —Ç–µ–∫—Å—Ç, —Ç–æ–ª—å–∫–æ —É–Ω–∏–∫–∞–ª—å–Ω—ã–µ –∑–Ω–∞—á–µ–Ω–∏—è
1. *perfect match sgRNA* - —Ç–µ–∫—Å—Ç
2. *gene* - —Ç–µ–∫—Å—Ç
3. *sgRNA sequence*, *genome input*, *sgRNA input* - —Ç–µ–∫—Å—Ç–æ–≤—ã–µ, —Å–æ–¥–µ—Ä–∂–∞—Ç –ø–æ—Å–ª–µ–¥–æ–≤–∞—Ç–µ–ª—å–Ω–æ—Å—Ç–∏ —Ç–æ–ª—å–∫–æ –∏–∑ –±—É–∫–≤ ATGC
4. *mismatch position* - –æ—Ç—Ä–∏—Ü–∞—Ç–µ–ª—å–Ω–æ–µ —Ü–µ–ª–æ–µ —á–∏—Å–ª–æ
5. *new pairing* - —Ç–µ–∫—Å—Ç
6. *K562*, *Jurkat* - –ª–æ–≥–∏—á–µ—Å–∫–∏–π —Ç–∏–ø
7. *mean relative gamma* - —á–∏—Å–ª–æ —Å –ø–ª–∞–≤–∞—é—â–µ–π —Ç–æ—á–∫–æ–π

### –°–æ–≥–ª–∞—Å–æ–≤–∞–Ω–Ω–æ—Å—Ç—å

**–£–Ω–∏–∫–∞–ª—å–Ω–æ—Å—Ç—å** - –Ω–∏ –æ–¥–∏–Ω –æ–±—ä–µ–∫—Ç –Ω–µ —Å—É—â–µ—Å—Ç–≤—É–µ—Ç –≤ –Ω–∞–±–æ—Ä–µ –¥–∞–Ω–Ω—ã—Ö –±–æ–ª–µ–µ –æ–¥–Ω–æ–≥–æ —Ä–∞–∑–∞. –ù–∞–ª–∏—á–∏–µ –¥—É–±–ª–µ–π –º–æ–∂–µ—Ç –ø—Ä–∏–≤–æ–¥–∏—Ç—å –∫ –Ω–µ—Å–æ–≥–ª–∞—Å–æ–≤–∞–Ω–Ω–æ—Å—Ç–∏ –∏ –ø—Ä–æ—Ç–∏–≤–æ—Ä–µ—á–∏—è–º –≤—Å–ª–µ–¥—Å—Ç–≤–∏–µ –æ—Ç—Å—É—Ç—Å—Ç–≤–∏—è –µ–¥–∏–Ω–æ–π –≤–µ—Ä—Å–∏–∏ –ø—Ä–∞–≤–¥—ã.

**–°–æ–≥–ª–∞—Å–æ–≤–∞–Ω–Ω–æ—Å—Ç—å** - —Å–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤–∏–µ –¥–∞–Ω–Ω—ã—Ö –¥—Ä—É–≥ –¥—Ä—É–≥—É –∏ –∏—Ö –ª–æ–≥–∏—á–µ—Å–∫–∞—è –Ω–µ–ø—Ä–æ—Ç–∏–≤–æ—Ä–µ—á–∏–≤–æ—Å—Ç—å.

1. –ó–Ω–∞—á–µ–Ω–∏—è *mismatch position* –∏ *new pairing* –¥–æ–ª–∂–Ω—ã —Å–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤–æ–≤–∞—Ç—å —Ä–µ–∞–ª—å–Ω–æ–º—É —Ä–∞–∑–ª–∏—á–∏—é –Ω—É–∫–ª–µ–æ—Ç–∏–¥–æ–≤ –º–µ–∂–¥—É *genome input* –∏ *sgRNA input*

## –†–∞–∑—Ä–∞–±–æ—Ç–∫–∞ –ë–î
–î–ª—è —Ö—Ä–∞–Ω–µ–Ω–∏—è –¥–∞–Ω–Ω—ã—Ö –±—ã–ª–∞ –≤—ã–±—Ä–∞–Ω–∞ –°–£–ë–î SQLite. –î–∞–Ω–Ω—ã–µ –∏–∑ txt-—Ñ–∞–π–ª–∞ –∏ –∏–∑ –¥–∞—Ç–∞—Ñ—Ä–µ–π–º–∞ –ø–µ—Ä–µ–Ω–æ—Å—è—Ç—Å—è –≤ –±–∞–∑—É –ø–æ—Å—Ä–µ–¥—Å—Ç–≤–æ–º —Å–∫—Ä–∏–ø—Ç–∞, –∏—Å–ø–æ–ª—å–∑—É—é—â–µ–≥–æ –≤—Å—Ç—Ä–æ–µ–Ω–Ω—ã–π –º–æ–¥—É–ª—å sqlite3.
### –°—Ç—Ä—É–∫—Ç—É—Ä–∞ —Ç–∞–±–ª–∏—Ü—ã clean_data
| –ù–∞–∑–≤–∞–Ω–∏–µ –ø–æ–ª—è         | –¢–∏–ø –¥–∞–Ω–Ω—ã—Ö | –û–≥—Ä–∞–Ω–∏—á–µ–Ω–∏—è                           |
|-----------------------|------------|---------------------------------------|
| `key`                 | TEXT       | PRIMARY KEY                           |
| `perfect_match_sgRNA` | TEXT       | NOT NULL                              |
| `gene`                | TEXT       |                                       |
| `sgRNA_sequence`      | TEXT       |                                       |
| `mismatch_position`   | INTEGER    | NOT NULL, CHECK(mismatch_position < 0)|
| `new_pairing`         | TEXT       |                                       |
| `K562`                | INTEGER    | NOT NULL, CHECK(K562 IN (0,1))        |
| `Jurkat`              | INTEGER    | NOT NULL, CHECK(Jurkat IN (0,1))      |
| `mean_relative_gamma` | REAL       | NOT NULL                              |
| `genome_input`        | TEXT       |                                       |
| `sgRNA_input`         | TEXT       |                                       |
| `encoded_or`          | TEXT       |                                       |
| `encoded_stacked`     | TEXT       |                                       |
| `encoded_7channels`   | TEXT       |                                       |
| `gc_content`          | REAL       |                                       |
| `pam`                 | TEXT       |                                       |

## –î–∞—à–±–æ—Ä–¥
–ò–Ω—Ç–µ—Ä–∞–∫—Ç–∏–≤–Ω—ã–π –¥–∞—à–±–æ—Ä–¥ —Ñ–æ—Ä–º–∏—Ä—É–µ—Ç—Å—è —Å –ø–æ–º–æ—â—å—é –±–∏–ª–∏–æ—Ç–µ–∫–∏ Streamlit (—Å–º. –ø. –ó–∞–ø—É—Å–∫ –¥–∞—à–±–æ—Ä–¥–∞)

### –ò—Å—Ç–æ—á–Ω–∏–∫–∏
1. –ö—Ä–∏—Ç–µ—Ä–∏–∏ –∫–∞—á–µ—Å—Ç–≤–∞ –¥–∞–Ω–Ω—ã—Ö / Loginom https://loginom.ru/blog/data-quality-criteria
2. –û—Ü–µ–Ω–∫–∞ –∫–∞—á–µ—Å—Ç–≤–∞ –¥–∞–Ω–Ω—ã—Ö / –Ø–Ω–¥–µ–∫—Å –û–±—Ä–∞–∑–æ–≤–∞–Ω–∏–µ https://education.yandex.ru/handbook/data-analysis/article/ocenka-kachestva-dannyh

## –ó–∞–ø—É—Å–∫

### –ü—Ä–µ–¥–≤–∞—Ä–∏—Ç–µ–ª—å–Ω—ã–µ —Ç—Ä–µ–±–æ–≤–∞–Ω–∏—è

- **Python 3.7** –∏–ª–∏ –≤—ã—à–µ.
- **–î–æ—Å—Ç—É–ø –≤ –∏–Ω—Ç–µ—Ä–Ω–µ—Ç** –¥–ª—è —Å–∫–∞—á–∏–≤–∞–Ω–∏—è –¥–∞–Ω–Ω—ã—Ö.

### –£—Å—Ç–∞–Ω–æ–≤–∫–∞

1. **–ö–ª–æ–Ω–∏—Ä—É–π—Ç–µ —Ä–µ–ø–æ–∑–∏—Ç–æ—Ä–∏–π** (–µ—Å–ª–∏ –ø—Ä–∏–º–µ–Ω–∏–º–æ):

    ```bash
    git clone https://github.com/–≤–∞—à_—Ä–µ–ø–æ–∑–∏—Ç–æ—Ä–∏–π.git
    cd –≤–∞—à_—Ä–µ–ø–æ–∑–∏—Ç–æ—Ä–∏–π
    ```

2. **–°–æ–∑–¥–∞–π—Ç–µ –∏ –∞–∫—Ç–∏–≤–∏—Ä—É–π—Ç–µ –≤–∏—Ä—Ç—É–∞–ª—å–Ω–æ–µ –æ–∫—Ä—É–∂–µ–Ω–∏–µ** (—Ä–µ–∫–æ–º–µ–Ω–¥—É–µ—Ç—Å—è):

    ```bash
    python -m venv venv
    source venv/bin/activate  # –î–ª—è Linux/macOS
    venv\Scripts\activate     # –î–ª—è Windows
    ```

3. **–£—Å—Ç–∞–Ω–æ–≤–∏—Ç–µ –Ω–µ–æ–±—Ö–æ–¥–∏–º—ã–µ –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏**:

    ```bash
    pip install -r requirements.txt
    ```

### –ó–∞–ø—É—Å–∫ –ø–∞–π–ø–ª–∞–π–Ω–∞

**–ò—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã—Ö –∑–Ω–∞—á–µ–Ω–∏–π**
```bash
python main.py
```
**–ò—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–∏—Ö –∑–Ω–∞—á–µ–Ω–∏–π**
```bash
python main.py --url "https://example.com/data.txt" --local_filename "data.txt" --db_name "my_database.db"
```

#### –û–ø–∏—Å–∞–Ω–∏–µ –ø–∞—Ä–∞–º–µ—Ç—Ä–æ–≤
+ `--url`:
  + *–û–ø–∏—Å–∞–Ω–∏–µ*: –°—Å—ã–ª–∫–∞ –Ω–∞ –∏—Å—Ö–æ–¥–Ω—ã–π —Ñ–∞–π–ª –¥–∞–Ω–Ω—ã—Ö, –∫–æ—Ç–æ—Ä—ã–π –Ω–µ–æ–±—Ö–æ–¥–∏–º–æ —Å–∫–∞—á–∞—Ç—å.
  + *–¢–∏–ø*: str
  + *–ü–æ —É–º–æ–ª—á–∞–Ω–∏—é*: https://raw.githubusercontent.com/ew314/sgRNA_seq2seq/main/relative_activity_predictor/data/Table_S8_machine_learning_input.txt

+ `--local_filename`:
  + *–û–ø–∏—Å–∞–Ω–∏–µ*: –ò–º—è –ª–æ–∫–∞–ª—å–Ω–æ–≥–æ —Ñ–∞–π–ª–∞, –≤ –∫–æ—Ç–æ—Ä—ã–π –±—É–¥—É—Ç —Å–æ—Ö—Ä–∞–Ω–µ–Ω—ã —Å–∫–∞—á–∞–Ω–Ω—ã–µ –¥–∞–Ω–Ω—ã–µ.
  + *–¢–∏–ø*: str
  + *–ü–æ —É–º–æ–ª—á–∞–Ω–∏—é*: Table_S8_machine_learning_input.txt

+ `--db_name`:
  + *–û–ø–∏—Å–∞–Ω–∏–µ*: –ò–º—è –±–∞–∑—ã –¥–∞–Ω–Ω—ã—Ö SQLite, –≤ –∫–æ—Ç–æ—Ä—É—é –±—É–¥—É—Ç –∑–∞–≥—Ä—É–∂–µ–Ω—ã –∏ —Å–æ—Ö—Ä–∞–Ω–µ–Ω—ã –¥–∞–Ω–Ω—ã–µ.
  + *–¢–∏–ø*: str
  + *–ü–æ —É–º–æ–ª—á–∞–Ω–∏—é*: crispr_sgRNA.db

### –ó–∞–ø—É—Å–∫ –¥–∞—à–±–æ—Ä–¥–∞
**–ò—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã—Ö –∑–Ω–∞—á–µ–Ω–∏–π**

```bash
streamlit run streamlit_app.py
```
**–ò—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–∏—Ö –∑–Ω–∞—á–µ–Ω–∏–π**
```bash
streamlit run streamlit_app.py -- --db_name=new_db.db
```

#### –û–ø–∏—Å–∞–Ω–∏–µ –ø–∞—Ä–∞–º–µ—Ç—Ä–æ–≤
+ `--db_name`:
  + *–û–ø–∏—Å–∞–Ω–∏–µ*: –ò–º—è –±–∞–∑—ã –¥–∞–Ω–Ω—ã—Ö SQLite, –∏–∑ –∫–æ—Ç–æ—Ä–æ–π –±—É–¥—É—Ç –≤—ã–≥—Ä—É–∂–µ–Ω—ã –¥–∞–Ω–Ω—ã–µ.
  + *–¢–∏–ø*: str
  + *–ü–æ —É–º–æ–ª—á–∞–Ω–∏—é*: crispr_sgRNA.db